ddl/foreign_key.go

```diff
@@ -112,3 +114,143 @@ func onDropForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ err
 		return ver, dbterror.ErrInvalidDDLState.GenWithStackByArgs("foreign key", fkInfo.State)
 	}
 }
+
+func allocateFKIndexID(tblInfo *model.TableInfo) int64 {
+	tblInfo.MaxForeignKeyID++
+	return tblInfo.MaxForeignKeyID
+}
+
+func checkTableForeignKeysValid(sctx sessionctx.Context, is infoschema.InfoSchema, schema string, tbInfo *model.TableInfo) error {
+	if !sctx.GetSessionVars().EnableForeignKey {
+		return nil
+	}
+	fkCheck := sctx.GetSessionVars().ForeignKeyChecks
+	for _, fk := range tbInfo.ForeignKeys {
+		if fk.Version < 1 {
+			continue
+		}
+		err := checkTableForeignKeyValid(is, schema, tbInfo, fk, fkCheck)
+		if err != nil {
+			return err
+		}
+	}
+
+	referredFKInfos := is.GetTableReferredForeignKeys(schema, tbInfo.Name.L)
+	for _, referredFK := range referredFKInfos {
+		childTable, err := is.TableByName(referredFK.ChildSchema, referredFK.ChildTable)
+		if err != nil {
+			return err
+		}
+		fk := model.FindFKInfoByName(childTable.Meta().ForeignKeys, referredFK.ChildFKName.L)
+		if fk == nil {
+			continue
+		}
+		err = checkTableForeignKey(tbInfo, childTable.Meta(), fk)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func checkTableForeignKeyValid(is infoschema.InfoSchema, schema string, tbInfo *model.TableInfo, fk *model.FKInfo, fkCheck bool) error {
+	var referTblInfo *model.TableInfo
+	if fk.RefSchema.L == schema && fk.RefTable.L == tbInfo.Name.L {
+		same := true
+		for i, col := range fk.Cols {
+			if col.L != fk.RefCols[i].L {
+				same = false
+				break
+			}
+		}
+		if same {
+			// self-reference with same columns is not support.
+			return infoschema.ErrCannotAddForeign
+		}
+		referTblInfo = tbInfo
+	} else {
+		referTable, err := is.TableByName(fk.RefSchema, fk.RefTable)
+		if err != nil {
```

Comment:
If the refer table is a local/global temporary table, should it report an error?

Reply:
nice catch, I think we can not support local/global temporary now.